### PR TITLE
Use GraphQL Yoga

### PR DESCRIPTION
**Pull request by ardatan**
Hi!
As we discussed recently. We decided to create this PR to migrate your GraphQL Server implementation to GraphQL Yoga.
GraphQL Yoga is platform agnostic GraphQL Server for JavaScript environment that follows Web standards and supports GraphQL-over-HTTP spec together with GraphQL Multipart spec, defer/stream and SSE subscriptions.
https://graphql-yoga.com/
Thanks to Yoga, you don't need any extra configuration and code for Envelop integration, handling and parsing requests, CORS and healthchecks.
Also we saw that you're using node-fetch already for polyfilling Node and abstracting some HTTP stuff. This PR replaces it with cross-undici-fetch which is a smart ponyfill library like node-fetch but chooses the best fetch implementation according to your Node version etc(node-fetch for v14,undici for v16, native fetch for v17 etc).

As far as I see, there is no change in the user facing API (not sure if onHealthCheck is something for users) so in the docs.

We believe you can leverage more from Fetch API like we do in GraphQL Yoga so GraphQL Yoga can have tighter integration. We'd appreciate your feedback. Thanks!

### CODE REVIEW

File: packages/api/src/cors.ts [hunk 1]
```diff
@@ -62,9 +63,7 @@ export function createCorsContext(cors: CorsConfig | undefined) {
       return request.method === 'OPTIONS'
     },
     getRequestHeaders(request: Request): CorsHeaders {
-      const eventHeaders = new Headers(
-        request.headers as Record<string, string>
-      )
+      const eventHeaders = new Headers(request.headers as HeadersInit)
```
[the remark this answers is not in the document]
Comment (ardatan) [reply packages/api/src/cors.ts:66]: No it is not typo. Request.headers can be an object map with the string arrays as values or Headers as well.

[the remark this answers is not in the document]
Comment (dac09) [reply packages/api/src/cors.ts:66]: Oh but where does it come from? Just available globally?

**Comment by dac09** [comment]
@ardatan thank you for this and the crazy quick turn around. The Guild never fails to impress me ♥️

File: packages/api/src/cors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ardatan) [reply packages/api/src/cors.ts:66]: Yes it is a globally available type

File: .github/renovate.json [hunk 2]
```diff
@@ -25,7 +25,6 @@
     "@types/node-fetch",
     "chalk",
     "pascalcase",
```
[the remark this answers is not in the document]
Comment (thedavidprice) [reply .github/renovate.json:26]: Ah, my mistake here. Searched through the PR and saw node-fetch removed everywhere. I also removed @types/node-fetch and removed both from renovate blocklist

**Comment by thedavidprice** [comment]
Updated from main to include new CI Auth Smoke Tests via #4673

File: packages/graphql-server/src/functions/graphql.ts [hunk 3]
```diff
@@ -182,52 +169,55 @@ export const createGraphQLHandler = ({
...
   })
 
   const handlerFn = async (
     event: APIGatewayProxyEvent,
     lambdaContext: LambdaContext
   ): Promise<APIGatewayProxyResult> => {
-    const enveloped = createSharedEnvelop({
-      event,
-      requestContext: lambdaContext,
+    const request = new Request(`http://localhost${event.path}`, {
```
[the remark this answers is not in the document]
Comment (dthyresson) [reply packages/graphql-server/src/functions/graphql.ts:212]: @ardatan and @dac09 is `localhost` here expected?

**Review by dthyresson — CHANGES_REQUESTED** [review]
Had one question about the use of `localhost` in the `hanlderFn`.

**Comment by dthyresson** [comment]
@dac09 and @ardatan I resolved the conflicts an updated the branch. I still think we have a few questions about use of localhost and perhaps CORS to think about before merge and testing.

**Comment by dthyresson** [comment]
I just realized I may have added back node-fetch as part of conflict resolution

File: packages/graphql-server/src/functions/graphql.ts [hunk 4]
```diff
@@ -182,52 +169,55 @@ export const createGraphQLHandler = ({
...
+        }
+
+        const requestAccessControlRequestHeaders = request.headers.get(
+          'access-control-request-headers'
+        )
+        if (!cors.allowedHeaders && requestAccessControlRequestHeaders) {
+          corsOptions.allowedHeaders = [requestAccessControlRequestHeaders]
+        }
+      }
+      return {}
```
[the remark this answers is not in the document]
Comment (ardatan) [reply packages/graphql-server/src/functions/graphql.ts:204]: Actually Envelop isn't responsible of CORS. So not sure if I understood what you mean there :) Yoga handles CORS and health check out of box. `corsOptions` is what Yoga generates by using `cors` object provided in the `createGraphQLHandler`'s parameters.

File: packages/graphql-server/src/functions/graphql.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ardatan) [reply packages/graphql-server/src/functions/graphql.ts:212]: `localhost` is just a dummy value there. We create a Request object for GraphQL Yoga then Yoga processes it and returns `Response`.

File: packages/graphql-server/src/functions/graphql.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ardatan) [reply packages/graphql-server/src/functions/graphql.ts:204]: Oh I just saw that I didn't use the same logic there. I'm fixing it now.

[the remark this answers is not in the document]
Comment (ardatan) [reply packages/graphql-server/src/functions/graphql.ts:235]: Oh my bad :)

File: packages/graphql-server/src/functions/graphql.ts [hunk 5]
```diff
@@ -182,52 +169,86 @@ export const createGraphQLHandler = ({
...
-  plugins.push(useMaskedErrors({ formatError, errorMessage: defaultError }))
+  const yoga = createServer({
+    schema,
+    plugins,
+    maskedErrors: {
+      formatError,
+      errorMessage: defaultError,
+    },
+    logging: logger,
+    graphiql: false,
```
[the remark this answers is not in the document]
Comment (dac09) [reply packages/graphql-server/src/functions/graphql.ts:180]: Yup yup... just trying it out now. Let me push up my changes in 2 mins!

[the remark this answers is not in the document]
Comment (dthyresson) [reply packages/graphql-server/src/functions/graphql.ts:180]: FYI old code was:

```ts
    if (isDevEnv && shouldRenderGraphiQL(request)) {
      return {
        body: renderPlaygroundPage({
          endpoint: graphiQLEndpoint || '/graphql',
        }),
        statusCode: 200,
        headers: {
          'Content-Type': 'text/html',
          ...corsHeaders,
        },
      }
    }
```

So, we made sure that we were in Dev and also that for headers got passed

[the remark this answers is not in the document]
Comment (ardatan) [reply packages/graphql-server/src/functions/graphql.ts:180]: Why do we need CORS headers for GraphiQL?

**Review by dthyresson — APPROVED** [review]

**Comment by dthyresson** [comment]
GraphQL Playground running on GitPod

image

**Review by dac09 — APPROVED** [review]

**Comment by borekb** [comment]
Just to confirm, this entirely replaces Helix, right? Helix is still mentioned in places like [README](https://github.com/redwoodjs/redwood/blob/30e9e6305061bbbb24f980811b5d448eca0d8b77/README.md#technologies) and a [few others](https://cs.github.com/redwoodjs/redwood?q=helix) but that might just be something that yet needs updating.

BTW were there any issues with Helix or is Yoga just easier / more convenient to use? We'll be doing a similar decision soon 😊.

**Comment by dthyresson** [comment]
Hi @borekb 
> Just to confirm, this entirely replaces Helix, right? 

Correct, the RedwoodJS GraphQL Server uses Yoga 2.0 now instead of Helix.

Helix is still mentioned in places like [README](https://github.com/redwoodjs/redwood/blob/30e9e6305061bbbb24f980811b5d448eca0d8b77/README.md#technologies) and a [few others](https://cs.github.com/redwoodjs/redwood?q=helix) but that might just be something that yet needs updating.

Thanks! We do need to clean up those mentions. Appreciate you finding them.

> BTW were there any issues with Helix or is Yoga just easier / more convenient to use? 

No, no issues with Helix. RedwoodJS decided to use Yoga based on the collaboration with the Guild.

While @dotansimha can explain more, in short, (my take) Yoga takes the "best parts" of Helix, adds in what the Guild had learned from their customers using it in production, better integrates their envelop plugin system, and is now a project that they can maintain, support and improve faster and more easily than Helix (which is maintained by another party).
 
> We'll be doing a similar decision soon

Sounds great. The transition for Redwood was really smooth.

**Pull request merged**